modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -610,178 +568,95 @@ private PreparedConfigurations prepareServiceConfigurations(Collection<ServiceCo
     public IgniteInternalFuture<?> deployAll(ClusterGroup prj, Collection<ServiceConfiguration> cfgs) {
         if (prj == null)
             // Deploy to servers by default if no projection specified.
-            return deployAll(cfgs,  ctx.cluster().get().forServers().predicate());
+            return deployAll(cfgs, ctx.cluster().get().forServers().predicate());
         else if (prj.predicate() == F.<ClusterNode>alwaysTrue())
-            return deployAll(cfgs,  null);
+            return deployAll(cfgs, null);
         else
             // Deploy to predicate nodes by default.
-            return deployAll(cfgs,  prj.predicate());
+            return deployAll(cfgs, prj.predicate());
     }
 
     /**
      * @param cfgs Service configurations.
      * @param dfltNodeFilter Default NodeFilter.
      * @return Future for deployment.
      */
-    private IgniteInternalFuture<?> deployAll(Collection<ServiceConfiguration> cfgs,
+    private IgniteInternalFuture<?> deployAll(@NotNull Collection<ServiceConfiguration> cfgs,
         @Nullable IgnitePredicate<ClusterNode> dfltNodeFilter) {
-        assert cfgs != null;
-
-        PreparedConfigurations srvCfg = prepareServiceConfigurations(cfgs, dfltNodeFilter);
-
-        List<ServiceConfiguration> cfgsCp = srvCfg.cfgs;
+        opsLock.readLock().lock();
```

Comment:
Why not `enterBusy()`?

Reply:
In case of using `enterBusy()` we should duplicate the checks below because of returning value of the method.
The main reason to use lock here to synchronize method which changing state (onDisconnect, onKernalStop), that means we should throw several types of exceptions.